src/sage/schemes/elliptic_curves/mod_poly.py

```diff
@@ -0,0 +1,140 @@
+
+from sage.databases.db_modular_polynomials import ClassicalModularPolynomialDatabase
+_db = ClassicalModularPolynomialDatabase()
+
+_cache = dict()
+
+def classical_modular_polynomial(l, j=None):
+    r"""
+    Return the classical modular polynomial `\Phi_\ell`, either as a
+    "generic" bivariate polynomial over `\ZZ`, or as an "instantiated"
+    modular polynomial where one variable has been replaced by the
+    given `j`-invariant.
+
+    Generic polynomials are cached up to a certain size of `\ell`,
+    which significantly accelerates subsequent invocations with the
+    same `\ell`. The default bound is `\ell \leq 150`, which can be
+    adjusted by setting ``classical_modular_polynomial.cache_bound``
+    to a different value. Beware that modular polynomials are very
+    large and the amount of memory consumed by the cache will grow
+    rapidly when the bound is set to a large value.
+
+    INPUT:
+
+    - ``l`` -- positive integer.
+
+    - ``j`` -- either ``None`` or a ring element.
+
+      - If ``None`` is given, the original modular polynomial
+        is returned as an element of `\ZZ[X,Y]`.
+
+      - If a ring element `j \in R` is given, the evaluation
+        `\Phi_\ell(j,Y)` is returned as an element of the
+        univariate polynomial ring `R[Y]`.
+
+    ALGORITHMS:
+
+    - The Kohel database
+      :class:`~sage.databases.db_modular_polynomials.ClassicalModularPolynomialDatabase`
+
+    - :pari:`polmodular`
+
+    EXAMPLES::
+
+        sage: classical_modular_polynomial(2)
+        -X^2*Y^2 + X^3 + 1488*X^2*Y + 1488*X*Y^2 + Y^3 - 162000*X^2 + 40773375*X*Y - 162000*Y^2 + 8748000000*X + 8748000000*Y - 157464000000000
+        sage: j = Mod(1728, 419)
+        sage: classical_modular_polynomial(3, j)
+        Y^4 + 230*Y^3 + 84*Y^2 + 118*Y + 329
+
+    TESTS::
+
+        sage: q = random_prime(50)^randrange(1,4)
+        sage: j = GF(q).random_element()
+        sage: l = random_prime(50)
+        sage: Y = polygen(parent(j), 'Y')
+        sage: classical_modular_polynomial(l,j) == classical_modular_polynomial(l)(j,Y)
+        True
+    """
+    l = ZZ(l)
+
+    if j is None:
+        # We are supposed to return the generic modular polynomial. First
+        # check if it is already in the cache, then check the database,
+        # finally compute it using PARI.
+        try:
+            return _cache[l]
+        except KeyError:
+            pass
+
+        try:
+            Phi = ZZ['X,Y'](_db[l])
+        except ValueError:
+            try:
+                pari_Phi = pari.polmodular(l)
+            except PariError:
+                raise NotImplementedError('modular polynomial is not in database and computing it on the fly is not yet implemented')
+            d = {(i,j): c for i,f in enumerate(pari_Phi) for j,c in enumerate(f)}
+            Phi = ZZ['X,Y'](d)
+
+        if l <= classical_modular_polynomial.cache_bound:
+            _cache[l] = Phi
+
+        return Phi
+
+    R,Y = parent(j)['Y'].objgen()
```

Comment:
```suggestion
    R = parent()['Y']
    Y = R.gen()
```
I find this to be easier to understand.